File: packages/node/motia/src/types-stream.ts
import { StreamAuthResult, StreamContext, StreamJoinResult } from '@iii-dev/sdk'
import type { FlowContext, StepSchemaInput } from './types'

export type StreamSubscription = { groupId: string; id?: string }

export interface StreamAuthInput {
  headers: Record<string, string>
  path: string
  queryParams: Record<string, string[]>
  addr: string
}

export type AuthenticateStream = (input: StreamAuthInput, context: FlowContext) => Promise<StreamAuthResult>

type PromiseOrValue<T> = T | Promise<T>

export interface StreamConfig {
  name: string
  schema: StepSchemaInput
  baseConfig: { storageType: 'default' } | { storageType: 'custom'; factory: () => MotiaStream<any> }
  /**
   * @deprecated Use onJoin instead
   */
  canAccess?: (subscription: StreamSubscription, authContext: StreamContext) => boolean | Promise<boolean>

  onJoin?: (
    subscription: StreamSubscription,
    context: FlowContext,
    authContext?: StreamContext,
  ) => PromiseOrValue<StreamJoinResult>
  onLeave?: (
    subscription: StreamSubscription,
    context: FlowContext,
    authContext?: StreamContext,
  ) => PromiseOrValue<void>
}

export type StateStreamEventChannel = { groupId: string; id?: string }
export type StateStreamEvent<TData> = { type: string; data: TData }

export type BaseStreamItem<TData = unknown> = TData & { id: string }

export type Stream<TConfig extends StreamConfig = StreamConfig> = {
  filePath: string
  config: TConfig
  hidden?: boolean
}

export interface MotiaStream<TData> {
  get(groupId: string, id: string): Promise<BaseStreamItem<TData> | null>
  set(groupId: string, id: string, data: TData): Promise<BaseStreamItem<TData>>
  delete(groupId: string, id: string): Promise<BaseStreamItem<TData> | null>
  getGroup(groupId: string): Promise<BaseStreamItem<TData>[]>

  send<T>(channel: StateStreamEventChannel, event: StateStreamEvent<T>): Promise<void>
}

File: frameworks/motia/motia-py/packages/motia/src/motia/types_stream.py
"""Stream type definitions."""

from typing import Any, Awaitable, Callable

from pydantic import BaseModel, ConfigDict

from .types import FlowContext


class StreamSubscription(BaseModel):
    """Stream subscription details."""

    group_id: str
    id: str | None = None


class StreamAuthInput(BaseModel):
    """Authentication input for stream connections."""

    headers: dict[str, str]
    path: str
    query_params: dict[str, list[str]]
    addr: str


class StreamAuthResult(BaseModel):
    """Authentication result for stream connections."""

    authorized: bool
    context: Any | None = None


class StreamConfig(BaseModel):
    """Configuration for a stream."""

    model_config = ConfigDict(arbitrary_types_allowed=True)

    name: str
    description: str | None = None
    schema: dict[str, Any] | None = None  # type: ignore[assignment]
    base_config: dict[str, Any] | None = None
    on_join: Callable[[StreamSubscription, FlowContext[Any], Any | None], Awaitable[Any]] | None = None
    on_leave: Callable[[StreamSubscription, FlowContext[Any], Any | None], Awaitable[Any]] | None = None
